src/v2i-hub/CARMACloudPlugin/src/CARMACloudPlugin.cpp

```diff
@@ -158,24 +211,88 @@ void CARMACloudPlugin::CARMAResponseHandler(QHttpEngine::Socket *socket)
 	tsm5message.set_contents(container.get_storage().get_tree());
 	tsm5ENC.encode_j2735_message(tsm5message);
 
-	msg.reset();
-	msg.reset(dynamic_cast<tsm5EncodedMessage*>(factory.NewMessage(api::MSGSUBTYPE_TESTMESSAGE05_STRING)));
-
-	string enc = tsm5ENC.get_encoding();
-	msg->refresh_timestamp();
-	msg->set_payload(tsm5ENC.get_payload_str());
-	msg->set_encoding(enc);
-	msg->set_flags(IvpMsgFlags_RouteDSRC);
-	msg->addDsrcMetadata(172, 0x8003);
-	msg->refresh_timestamp();
-
-	routeable_message *rMsg = dynamic_cast<routeable_message *>(msg.get());
-	BroadcastMessage(*rMsg);
-
-	PLOG(logERROR) << " CARMACloud Plugin :: Broadcast tsm5:: " << tsm5ENC.get_payload_str();
+	//Get TCM id
+	Id64b_t tcmv01_req_id = tsm5message.get_j2735_data()->body.choice.tcmV01.reqid;
+	
+	ss.str(""); 
+    for(size_t i=0; i < tcmv01_req_id.size; i++)
+    {
+		ss << std::setfill('0') << std::setw(2) << std::hex << (unsigned) tcmv01_req_id.buf[i];
+    }
+	string tcmv01_req_id_hex = ss.str();	
+	
+	std::transform(tcmv01_req_id_hex.begin(), tcmv01_req_id_hex.end(), tcmv01_req_id_hex.begin(), ::tolower );	
+	if(tcmv01_req_id_hex.length() > 0)
+	{
+		std::lock_guard<mutex> lock(_not_ACK_TCMs_mutex);
+		_not_ACK_TCMs->insert({tcmv01_req_id_hex, tsm5ENC});
+	}	
 }
 
-
+void CARMACloudPlugin::Broadcast_TCMs()
+{ 	
+	std::time_t start_time = 0, cur_time = 0;
+	bool is_started_broadcasting = false;
+	while(true)

```

Comment:
There is a sonar code smell associated with more than three nest while/if/for loops. Two ways we may be able to avoid this is either by using recursion : 
```
Broadcast_TCMs() {
// do something
if (broadcasting is not done)
{
Broadcast_TCMs()
}

```

Or we could move the while loop out side of the broadcast method and just make it a send method that is called repeatedly

Reply:
The method seems already does it